### PR TITLE
[KeyValueSnapshotReader Phase I] Part 8: ReadSentinelHandler... and fixing a subtle 🐛

### PR DESCRIPTION
**Pull request by jeremyk-91**
## General
**Before this PR**: The SnapshotTransaction class is responsible for handling too much, including management of bare read sentinels after a user data table has been manipulated non-atomically (e.g. truncation, or when restoring from backup).

There is also a bug with the original implementation, though it is probably benign given that this bug manifests very loudly when it is relevant. The current implementation for handling read sentinels basically acknowledges that bare sentinels, or sentinels only covered by uncommitted values, are fair game to be treated as empty. (Notice that this should never happen by the Atlas write protocol: you only put a sentinel when you delete something from sweep, and failed transactions are generally gotten rid of using direct deletes, not range tombstones). 

The code for handling this, however, used a null check on the return value of a TransactionService. In other words, suppose some cell has a bare sentinel. Now suppose that we attempt to write to that cell, but abort. All future reads of that version of that cell will read the aborted version, go through post filtering, see the sentinel, think that it is not a bare sentinel (because there's some timestamp -> -1 in the txn table) and panic. But it _is_ a bare sentinel, just that the code does not suggest as such!

**After this PR**:

==COMMIT_MSG==
SnapshotTransaction no longer throws when reading a bare sentinel covered by an aborted value, which can happen post-table truncation or restores if subsequent transactions fail.
==COMMIT_MSG==

Also SnapshotTransaction has fewer responsibilities, though the former is probably more changelog-relevant.

**Priority**: High P2

**Concerns / possible downsides (what feedback would you like?)**: Nothing in particular.

**Is documentation needed?**: No.

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**: Don't think so

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**: No

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**: Yes

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**: Not that I'm aware of

**Does this PR need a schema migration?** No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**: Nothing specific

**What was existing testing like? What have you done to improve it?**: Extended existing tests to use shiny JUnit5 parameterized tests for "how a value can be uncommitted", and wrote new unit tests for the small piece extracted.

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**: Not much concurrency...

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**: N/A

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**: Nothing breaks! The specific situation is so rare that I don't know if we can meaningfully test it.

**Has the safety of all log arguments been decided correctly?**: Enum types only.

**Will this change significantly affect our spending on metrics or logs?**: No.

**How would I tell that this PR does not work in production? (monitors, etc.)**: Something breaks

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**: Rollback

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**: No

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**: No

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**: Not that I'm aware of

## Development Process
**Where should we start reviewing?**: Either the AbstractSnapshotTransaction test changes or the ReadSentinelTest

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**: This could be done, but the PR isn't dramatically high priority, and a good chunk is just moved from SnapshotTransaction to the separate class.

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/DefaultOrphanedSentinelDeleter.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    @Override
+    public void scheduleSentinelsForDeletion(TableReference tableReference, Set<Cell> orphanedSentinels) {
+        if (orphanedSentinels.isEmpty()) {
+            return;
+        }
+        if (sweepStrategyProvider
+                .apply(tableReference)
+                .getSweeperStrategy()
+                .map(sweeperStrategy -> sweeperStrategy == SweeperStrategy.THOROUGH)
+                .orElse(false)) {
```
Comment (LucasIME) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/DefaultOrphanedSentinelDeleter.java]: nit: I particularly dislike long statements like this on if statements. I'd prefer if we extracted this to a boolean variable with a meaningful name like `isThoroughSweepStratety` and used that instead in the if statement.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 2]
```diff
@@ -390,6 +388,11 @@ private enum State {
         this.preCommitRequirementValidator = new DefaultPreCommitRequirementValidator(
                 preCommitCondition, transactionOutcomeMetrics, immutableTimestampLockManager);
         this.commitTimestampLoader = commitTimestampLoader;
+        this.readSentinelHandler = new ReadSentinelHandler(
+                transactionKeyValueService,
+                transactionService,
+                readSentinelBehavior,
+                new DefaultOrphanedSentinelDeleter(sweepStrategyManager::get, deleteExecutor));
```
Comment (LucasIME) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:395]: more of a question than a comment: along these refactors we've been doing a bunch of object instantiations here, rather the usual preferred method of dependency injection through the constructor. I assume that it's intentional so we don't break the constructor and just introduce a new one where we pass all of these once we've stabilized the interface at the end of the refactors?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadSentinelHandler.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            }
+
+            Map<Cell, Long> cellsToQuery = sentinelTypeToTimestamps.get(SentinelType.INDETERMINATE);
+            if (cellsToQuery == null) {
+                break;
+            }
+            Set<Long> committedStartTimestamps = KeyedStream.stream(transactionService.get(cellsToQuery.values()))
+                    .filter(commit -> commit != null && commit != TransactionConstants.FAILED_COMMIT_TS)
+                    .keys()
+                    .collect(Collectors.toSet());
```
Comment (LucasIME) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadSentinelHandler.java:98]: feedback for next time: would've been useful to leave a message here highlighting the different between the new and old version and where we're fixing the bug.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/AbstractSnapshotTransactionTest.java [hunk 4]
```diff
@@ -3650,4 +3684,9 @@ private static List<Cell> generateCells(int numberOfRowsPerTable, int numberOfCe
                         Cell.create(PtBytes.toBytes("row" + (i % numberOfCellsPerRow)), PtBytes.toBytes("column" + i)))
                 .collect(Collectors.toUnmodifiableList());
     }
+
+    private enum UncommittedValueType {
```
Comment (LucasIME) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/AbstractSnapshotTransactionTest.java]: might as well extract to a separate test class file if we're using in more than one test class?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:395]: Yep, that's right. Not the long-term goal to have too much construction logic here, indeed.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/AbstractSnapshotTransactionTest.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/AbstractSnapshotTransactionTest.java]: Yep that makes sense!

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadSentinelHandler.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadSentinelHandler.java:98]: yep! will do - yeah the difference is indeed that `commit != null` alone which was previously used, is insufficient

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/DefaultOrphanedSentinelDeleter.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/DefaultOrphanedSentinelDeleter.java]: OK, changed! Fair enough

**Review by LucasIME — APPROVED** [review]

**Pull request merged**